File: python/pytoon/encode/writer.py
from __future__ import annotations

from pytoon.types import Depth


class LineWriter:
    __slots__ = ('_lines', '_indent_unit')

    def __init__(self, indent_size: int) -> None:
        self._lines: list[str] = []
        self._indent_unit = ' ' * indent_size

    def push(self, depth: Depth, content: str) -> None:
        self._lines.append(f"{self._indent_unit * depth}{content}")

    def push_list_item(self, depth: Depth, content: str) -> None:
        from pytoon.constants import LIST_ITEM_PREFIX  # local import to avoid cycles

        self.push(depth, f"{LIST_ITEM_PREFIX}{content}")

    def to_string(self) -> str:
        return '\n'.join(self._lines)

File: packages/toon_julia/src/encode/writer.jl
"""
LineWriter handles building the output string with proper indentation.
"""
mutable struct LineWriter
    lines::Vector{String}
    indentation_string::String
    
    function LineWriter(indent_size::Int)
        new(String[], " " ^ indent_size)
    end
end

"""
Pushes a line with the specified depth (indentation level).
"""
function push(writer::LineWriter, depth::Int, content::String)
    indent = writer.indentation_string ^ depth
    push!(writer.lines, indent * content)
end

"""
Pushes a list item with the specified depth.
"""
function push_list_item(writer::LineWriter, depth::Int, content::String)
    push(writer, depth, LIST_ITEM_PREFIX * content)
end

"""
Converts the writer to a string.
"""
function Base.string(writer::LineWriter)::String
    return join(writer.lines, "\n")
end

